core/src/test/java/com/linecorp/armeria/common/RequestPeekTest.java

```diff
@@ -0,0 +1,149 @@
+        final HttpRequest req = HttpRequest.of(RequestHeaders.of(HttpMethod.GET, "/test"),
+                                               HttpData.ofUtf8("foo"));
+        final HttpRequest transformed = req
+                .peekData(data -> assertThat(data.toStringUtf8()).doesNotContain(" bar"))
+                .mapData(data -> HttpData.ofUtf8(data.toStringUtf8() + " bar"))
+                .peekData(data -> assertThat(data.toStringUtf8()).contains(" bar"));
+
+        assertThat(transformed.aggregate().join().contentUtf8()).isEqualTo("foo bar");
+        assertThat(transformed.headers()).isEqualTo(req.headers());
+    }
+
+    @Test
+    void peekTrailers() {
+        final HttpRequest noTrailers = HttpRequest.of(RequestHeaders.of(HttpMethod.GET, "/test"),
+                                                      HttpData.ofUtf8("foo"));
+        final AtomicBoolean invoked = new AtomicBoolean();
+        final HttpRequest transformed1 = noTrailers
+                .peekTrailers(trailers -> peekCount.incrementAndGet())
+                .mapTrailers(trailers -> {
+                    invoked.set(true);
+                    return trailers;
+                });
+
+        final AggregatedHttpRequest aggregated1 = transformed1.aggregate().join();
+        assertThat(invoked).isFalse();
+        assertThat(peekCount).hasValue(0);
+        assertThat(transformed1.headers()).isEqualTo(noTrailers.headers());
+        assertThat(aggregated1.contentUtf8()).isEqualTo("foo");
+        assertThat(aggregated1.trailers().isEmpty()).isTrue();
+
+        final HttpRequest withTrailers = HttpRequest.of(RequestHeaders.of(HttpMethod.GET, "/test"),
+                                                        HttpData.ofUtf8("foo"),
+                                                        HttpHeaders.of("trailer1", "1"));
+
+        final HttpRequest transformed2 = withTrailers
+                .peekTrailers(trailers -> {
+                    assertThat(trailers.get("trailer1")).isEqualTo("1");
+                    assertThat(trailers.get("trailer2")).isNull();
+                })
+                .mapTrailers(trailers -> {
+                    invoked.set(true);
+                    if ("1".equals(trailers.get("trailer1"))) {
+                        return trailers.toBuilder().add("trailer2", "2").build();
+                    }
+                    return trailers;
+                })
+                .peekTrailers(trailers -> {
+                    assertThat(trailers.get("trailer1")).isEqualTo("1");
+                    assertThat(trailers.get("trailer2")).isEqualTo("2");
+                })
+                .peekTrailers(trailers -> peekCount.incrementAndGet());
+
+        final AggregatedHttpRequest aggregated2 = transformed2.aggregate().join();
+        assertThat(invoked).isTrue();
+        assertThat(peekCount).hasValue(1);
+        assertThat(aggregated2.contentUtf8()).isEqualTo("foo");
+        assertThat(aggregated2.trailers().size()).isEqualTo(2);
```

Comment:
prefer `assertj` specific methods for more clear test failure messages

```suggestion
        assertThat(aggregated2.trailers()).hasSize(2);
```